horovod/common/operations.cc

```diff
@@ -148,6 +148,11 @@ OperationManager* CreateOperationManager(HorovodGlobalState& state) {
   std::vector<std::shared_ptr<BroadcastOp>> broadcast_ops;
   std::vector<std::shared_ptr<AllreduceOp>> adasum_ops;
 
+#if HAVE_NCCL && HOROVOD_GPU_ALLGATHER == 'N'
+  allgather_ops.push_back(std::shared_ptr<AllgatherOp>(
+      new NCCLAllgather(&nccl_context, &gpu_context, &state)));
+#endif
```

Comment:
I put it here because I think the priority of NCCLAllgather should higher than MPIHierarchicalAllgather. What do you think about this?

Reply:
The behavior should be that if a user sets `HOROVOD_HIERARCHICAL_ALLGATHER=1` that the hierarchical mode is used. Since there is no NCCL-based hierarchical allgather, this flag should trigger the use of `MPIHierarchicalAllgather`. For this to happen, the `NCCLAllgather` op has to be pushed to the vector after `MPIHierarchicalAllgather`. By default,  `HOROVOD_HIERARCHICAL_ALLGATHER=0` which disables `MPIHierarchicalAllgather`, so `NCCLAllgather` will be used in the default case.

Reply:
Thanks for clarifying. I have change it.